src/expr/src/expr/array.rs

```diff
@@ -0,0 +1,192 @@
+use std::sync::Arc;
+
+use itertools::Itertools;
+use risingwave_common::array::*;
+use risingwave_common::row::OwnedRow;
+use risingwave_common::types::to_text::ToText;
+use risingwave_common::types::{DataType, Datum};
+use risingwave_pb::expr::expr_node::{RexNode, Type};
+use risingwave_pb::expr::ExprNode;
+
+use crate::expr::{build_from_prost, Expression};
+use crate::{bail, ensure, ExprError, Result};
+
+/// Converts each array element to its text representation, and concatenates those
+/// separated by the delimiter string. If `null_string` is given and is not NULL,
+/// then NULL array entries are represented by that string; otherwise, they are omitted.
+///
+/// ```sql
+/// array_to_string ( array anyarray, delimiter text [, null_string text ] ) → text
+/// ```
+///
+/// Examples:
+///
+/// ```slt
+/// query T
+/// select array_to_string(array[1, 2, 3, NULL, 5], ',')
+/// ----
+/// 1,2,3,5
+///
+/// query T
+/// select array_to_string(array[1, 2, 3, NULL, 5], ',', '*')
+/// ----
+/// 1,2,3,*,5
+///
+/// query T
+/// select array_to_string(array[null,'foo',null], ',', '*');
+/// ----
+/// *,foo,*
+///
+/// query T
+/// with t as (
+///   select array[1,null,2,3] as arr, ',' as d union all
+///   select array[4,5,6,null,7] as arr, '|')
+/// select array_to_string(arr, d) from t;
+/// ----
+/// 1,2,3
+/// 4|5|6|7
+///
+/// # `array` or `delimiter` are required. Otherwise, returns null.
+/// query T
+/// select array_to_string(array[1,2], NULL);
+/// ----
+/// NULL
+///
+/// # FIXME: this query will panic in binder
```

Comment:
```
thread 'risingwave-main' panicked at 'Functions returning struct or list can not be inferred. Please use `FunctionCall::new_unchecked`.', src/common/src/types/mod.rs:189:47
stack backtrace:
   0: rust_begin_unwind
             at /rustc/3984bc5833db8bfb0acc522c9775383e4171f3de/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/3984bc5833db8bfb0acc522c9775383e4171f3de/library/core/src/panicking.rs:64:14
   2: <risingwave_common::types::DataType as core::convert::From<risingwave_common::types::DataTypeName>>::from::{{closure}}
             at ./src/common/src/types/mod.rs:189:47
   3: core::option::Option<T>::unwrap_or_else
             at /rustc/3984bc5833db8bfb0acc522c9775383e4171f3de/library/core/src/option.rs:873:21
   4: <risingwave_common::types::DataType as core::convert::From<risingwave_common::types::DataTypeName>>::from
             at ./src/common/src/types/mod.rs:189:9
   5: <T as core::convert::Into<U>>::into
             at /rustc/3984bc5833db8bfb0acc522c9775383e4171f3de/library/core/src/convert/mod.rs:726:9
   6: risingwave_frontend::expr::type_inference::func::infer_type::{{closure}}
             at ./src/frontend/src/expr/type_inference/func.rs:50:43
...
  20: risingwave_frontend::expr::type_inference::func::infer_type
             at ./src/frontend/src/expr/type_inference/func.rs:45:15
  21: risingwave_frontend::expr::function_call::FunctionCall::new
             at ./src/frontend/src/expr/function_call.rs:103:27
  22: risingwave_frontend::binder::expr::function::<impl risingwave_frontend::binder::Binder>::bind_builtin_scalar_function::rewrite::{{closure}}
             at ./src/frontend/src/binder/expr/function.rs:257:20
```

It can be fixed by **either**:

1. Implement the type coercion rule in `infer_type_for_special` instead of `src/expr/src/sig/func.rs`
2. Turn `impl From<DataTypeName> for DataType` to `TryFrom`

https://github.com/risingwavelabs/risingwave/blob/229fbbbdcbfabd3624b91a8f43f048ead4fad0ad/src/common/src/types/mod.rs#L187-L191

I prefer 2 because a `Try` implementation with potential panic looks weird... But this would lead to lots other unrelated changes. I'll try this in another PR.

cc. @xiangjinwu

Reply:
We can move to `TryFrom` but it does not solve the problem, and the caller still need to unwrap or use a branching statement like `infer_type_for_special`, because the general inferring logic cannot handle lists well:

When the user inputs `array_to_string(null, ',');`, what is the type of `null`? It could be `int[]`, or `date[]`, or `int[][][]`. The result of expr is `null::text` regardless of input type, but we do not have a placeholder like `anyarray` in PostgreSQL. So we have to check for it with special logic in frontend.

In PostgreSQL:
```
test=# select array_to_string(null, ',');
ERROR:  could not determine polymorphic type because input has type unknown
```

Reply:
>  because the general inferring logic cannot handle lists well

Yes, it's not supposed to be handled.

My idea was to solve the problem (panic) by returning an `Err` when `TryFrom` returns `Err`.

I was trying to avoid doing it with `infer_type_for_special` because that looks verbose. Since there are [many other array functions](https://www.postgresql.org/docs/current/functions-array.html), I hope to solve it elegantly.

Reply:
Fixed with a workaround. See: `src/frontend/src/expr/type_inference/func.rs`